### PR TITLE
Add a Markdown plugin and new syntax for rendering the Details element

### PR DESCRIPTION
**Pull request by Hamms**
This plugin adds a new Markdown syntax for rendering details and summary tags, a pattern we use extensively in our course content. The proposed syntax is inspired by [this proposal for a generic directives syntax](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444), and is intended to be sufficiently generic that we could in the future add more syntaxes matching this pattern.

Specifically, we implement a version of the **Container Block Directives** in the proposal, with the following changes:

1. The generic "name" tag must always be the exact string `"details"`
2. Generic attributes (key-value pairs) are not supported
3. The content block is required

For example, the following `details` element (from https://studio.code.org/s/csd2-2019/stage/3/puzzle/8):

```html
<details>

<summary>
<strong>What is a paragraph element?</strong>
</summary>

Paragraphs are marked by opening(`<p>`) and closing(`</p>`) tags. Paragraphs in HTML can be any length of text from one word to a bunch of sentences. Paragraphs group together sets of sentences and put some space between that group of text and the next group of text.

</details>
```

Could be written as:

```
::: details [**What is a paragraph element?**]
Paragraphs are marked by opening(`<p>`) and closing(`</p>`) tags. Paragraphs in HTML can be any length of text from one word to a bunch of sentences. Paragraphs group together sets of sentences and put some space between that group of text and the next group of text.
:::
```

Note that in the original, the summary text had to be bolded in HTML rather than with markdown syntax, and the body content had to be separated from the tags by newlines in order for the markdown syntax to be parsed correctly. With this new syntax, neither of those are necessary.

See the new unit test for more examples of what this syntax is capable of.

### CODE REVIEW

**Comment by bethanyaconnor** [comment]
Why do we need to redact the markdown here?

File: apps/test/unit/templates/plugins/detailsTest.js [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+      '<details><summary>summary-content</summary><p>contents, which are sometimes further block elements</p></details>\n';
+
+    const rendered = parser.sourceToHtml(markdown);
+    assert.equal(rendered, expected);
+  });
+
+  it('can nest', () => {
+    const markdown =
+      ':::: details [outer]\n' +
+      '::: details [inner]\n' +
```
Comment (bethanyaconnor) [inline apps/test/unit/templates/plugins/detailsTest.js:96]: Do nested details elements need to be proceeded by fewer colons than their parent(s)? If so, is that mentioned anywhere?

File: apps/test/unit/templates/plugins/detailsTest.js [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    const expected =
+      '<p>::: details\n' +
+      'contents, which are sometimes further block elements\n' +
+      ':::</p>\n';
+
+    const rendered = parser.sourceToHtml(markdown);
+    assert.equal(rendered, expected);
+  });
+
+  it('requires at least three opening colons', () => {
```
Comment (bethanyaconnor) [inline apps/test/unit/templates/plugins/detailsTest.js:121]: Is this requirement mentioned anywhere?

File: apps/test/unit/templates/plugins/detailsTest.js (hunk 1 above)
Comment (Hamms) [reply apps/test/unit/templates/plugins/detailsTest.js:96]: Sort of. Blocks are closed by a line of colons longer than or equal to the opening, so in order to make sure the line which closes the inner block only closes it and not the outer block, the inner block needs to be closed by a line that's too short to close the outer.

From [the original proposal](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444):

> Analogous to fenced code blocks, an arbitrary number of colons greater or equal three could be used as long as the closing line is longer than the opening line. That way, you can even nest blocks (think divs) by using successively fewer colons for each containing block.

File: apps/test/unit/templates/plugins/detailsTest.js (hunk 2 above)
Comment (Hamms) [reply apps/test/unit/templates/plugins/detailsTest.js:121]: This is also mentioned in the original proposal

**Comment by Hamms** [comment]
Our general philosophy around markdown is that as much syntax as possible should be redacted. We're even considering adding redaction for things like italics and bold; something like this that provides lots of opportunities for syntax to be accidentally broken by translators should _definitely_ be redacted.

File: apps/test/unit/templates/plugins/detailsTest.js (hunk 1 above)
Comment (bethanyaconnor) [reply apps/test/unit/templates/plugins/detailsTest.js:96]: Ah I figured I missed this somewhere. Thanks!

**Comment by bethanyaconnor** [comment]
Sounds good -- thanks for explanation!

**Review by bethanyaconnor — APPROVED** [review]

**Pull request merged**